File: python/math-and-geometry/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = {}
        while n != 1:
            if n in seen:
                return False

            sumofdigits = 0
            for char in str(n):
                sumofdigits += int(char) ** 2

            seen[n] = True
            n = sumofdigits

        return True

File: c++/math-and-geometry/happy-number.cc
class Solution {
public:
    bool isHappy(int n) {
        int num = n;
        std::unordered_set<int> visited;
        while (num != 1) {
            if (visited.contains(num)) {
                return false;
            }

            visited.insert(num);

            int new_num = 0;
            while (num > 0) {
                new_num += std::pow(num % 10, 2);
                num /= 10;
            }
            num = new_num;
        }
        return true;
    }
};
